### PR TITLE
Add cluster support to MariaDB

### PR DESCRIPTION
**Pull request by javsalgar**
**What this PR does / why we need it**:
Add master-slave support to MariaDB chart

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: stable/mariadb/README.md [hunk 1]
```diff
@@ -2,29 +2,29 @@
...
 
-MariaDB is developed as open source software and as a relational database, it provides an SQL interface for accessing data. The latest versions of MariaDB also include GIS and JSON features.
+MariaDB is developed as open source software and as a relational database it provides an SQL interface for accessing data. The latest versions of MariaDB also include GIS and JSON features.
 
-## TL;DR;
+## TL;DR
 
 ```bash
-$ helm install stable/mariadb
+$ helm install bitnami/mariadb-cluster
```
Comment (timstoop) [inline stable/mariadb/README.md]: This way, you're not getting this Chart...

File: stable/mariadb/README.md [hunk 2]
```diff
@@ -2,29 +2,29 @@
...
+- Kubernetes 1.4+ with Beta APIs enabled
 - PV provisioner support in the underlying infrastructure
 
 ## Installing the Chart
 
 To install the chart with the release name `my-release`:
 
 ```bash
-$ helm install --name my-release stable/mariadb
+$ helm install --name my-release bitnami/mariadb-cluster
```
Comment (timstoop) [inline stable/mariadb/README.md]: Here as well, you're telling the user to install from the bitname Chart repository, which is not where this Chart resides.

File: stable/mariadb/README.md [hunk 3]
```diff
@@ -43,122 +43,93 @@ The command removes all the Kubernetes components associated with the chart and
...
+| `db.user`                                 | Username of new user to create                      | `nil`                                                             |
+| `db.password`                             | Password for the new user                           | _random 10 character alphanumeric string if `db.user` is defined_ |
+| `db.name`                                 | Name for new database to create                     | `my_database`                                                     |
+| `replication.enabled`                     | MariaDB replication enabled                         | `true`                                                            |
+| `replication.user`                        | MariaDB replication user                            | `replicator`                                                      |
+| `replication.password`                    | MariaDB replication user password                   | _random 10 character alphanumeric string_                         |
+| `master.antiAffinity`                     | Master pod anti-affinity policy                     | `soft`                                                            |
```
Comment (timstoop) [inline stable/mariadb/README.md:64]: Please let the user set the entire affinity block (like the mongodb-replicaset Chart, but most other that allow affinity to be set as well) as it's more flexible.

File: stable/mariadb/README.md [hunk 4]
```diff
@@ -43,122 +43,93 @@ The command removes all the Kubernetes components associated with the chart and
...
+| `master.readinessProbe.initialDelaySeconds`| Delay before readiness probe is initiated (master) | `15`                                                              |
+| `master.readinessProbe.periodSeconds`     | How often to perform the probe (master)             | `10`                                                              |
+| `master.readinessProbe.timeoutSeconds`    | When the probe times out (master)                   | `1`                                                               |
+| `master.readinessProbe.successThreshold`  | Minimum consecutive successes for the probe (master)| `1`                                                               |
+| `master.readinessProbe.failureThreshold`  | Minimum consecutive failures for the probe (master) | `3`                                                               |
+| `slave.replicas`                          | Desired number of slave replicas                    | `1`                                                               |
+| `slave.antiAffinity`                      | Slave pod anti-affinity policy                      | `soft`                                                            |
```
Comment (timstoop) [inline stable/mariadb/README.md:85]: Like with the master, please allow the setting of the entire affinity block.

File: stable/mariadb/README.md [hunk 5]
```diff
@@ -43,122 +43,93 @@ The command removes all the Kubernetes components associated with the chart and
...
 The above parameters map to the env variables defined in [bitnami/mariadb](http://github.com/bitnami/bitnami-docker-mariadb). For more information please refer to the [bitnami/mariadb](http://github.com/bitnami/bitnami-docker-mariadb) image documentation.
 
 Specify each parameter using the `--set key=value[,key=value]` argument to `helm install`. For example,
 
 ```bash
 $ helm install --name my-release \
-  --set mariadbRootPassword=secretpassword,mariadbUser=my-user,mariadbPassword=my-password,mariadbDatabase=my-database \
-    stable/mariadb
+  --set root.password=secretpassword,user.database=app_database \
+    bitnami/mariadb-cluster
```
Comment (timstoop) [inline stable/mariadb/README.md]: Again, this should be `stable/mariadb`, as that's this repository.

File: stable/mariadb/README.md [hunk 6]
```diff
@@ -43,122 +43,93 @@ The command removes all the Kubernetes components associated with the chart and
...
 ```
 
-The above command sets the MariaDB `root` account password to `secretpassword`. Additionally, it creates a standard database user named `my-user`, with the password `my-password`, who has access to a database named `my-database`.
+The above command sets the MariaDB `root` account password to `secretpassword`. Additionally it creates a database named `my_database`.
 
 Alternatively, a YAML file that specifies the values for the parameters can be provided while installing the chart. For example,
 
 ```bash
-$ helm install --name my-release -f values.yaml stable/mariadb
+$ helm install --name my-release -f values.yaml bitnami/mariadb-cluster
```
Comment (timstoop) [inline stable/mariadb/README.md]: `stable/mariadb` is this Chart

File: stable/mariadb/templates/NOTES.txt [hunk 7]
```diff
@@ -1,66 +1,35 @@
...
-{{- end }}
-
-** Please be patient while the chart is being deployed **
-
-MariaDB can be accessed via port 3306 on the following DNS name from within your cluster:
-{{ template "mariadb.fullname" . }}.{{ .Release.Namespace }}.svc.cluster.local
-
-{{ if .Values.usePassword -}}
-To get the root password run:
+  Watch the deployment status using the command: kubectl get pods -w --namespace {{ .Release.Namespace }}
```
Comment (timstoop) [inline stable/mariadb/templates/NOTES.txt]: While you're at, also add `-l app=mariadb` to that commandline :)

File: stable/mariadb/templates/NOTES.txt [hunk 8]
```diff
@@ -1,66 +1,35 @@
...
 
-    export MARIADB_ROOT_PASSWORD=$(kubectl get secret --namespace {{ .Release.Namespace }} {{ template "mariadb.fullname" . }} -o jsonpath="{.data.mariadb-root-password}" | base64 --decode)
+Services:
 
-{{ if .Values.mariadbUser -}}
-To get the password for "{{ .Values.mariadbUser }}" run:
-
-    export MARIADB_PASSWORD=$(kubectl get secret --namespace {{ .Release.Namespace }} {{ template "mariadb.fullname" . }} -o jsonpath="{.data.mariadb-password}" | base64 --decode)
+  echo Master: {{ template "master.fullname" . }}.{{ .Release.Namespace }}.svc.cluster.local:{{ .Values.service.port }}
```
Comment (timstoop) [inline stable/mariadb/templates/NOTES.txt:10]: What's the `echo` for?

File: stable/mariadb/values.yaml [hunk 9]
```diff
@@ -1,110 +1,233 @@
...
-  image: prom/mysqld-exporter
-  imageTag: v0.10.0
-  imagePullPolicy: IfNotPresent
+  image:
+    registry: docker.io
+    repository: prom/mysqld-exporter
+    tag: v0.10.0
+    pullPolicy: IfNotPresent
   resources: {}
   annotations:
```
Comment (timstoop) [inline stable/mariadb/values.yaml:231]: This one is not in the README.

**Review by timstoop — CHANGES_REQUESTED** [review]
One thing that bugs me is that unless the bitnami image does some serious cool magic, you won't be able to get an additional slave up and running by simply increasing the number of replicas for the slaves. I think that should be very clearly explained in the README and probably the NOTES as well.

File: stable/mariadb/README.md (hunk 1 above)
Comment (javsalgar) [reply stable/mariadb/README.md]: Fixed

File: stable/mariadb/README.md (hunk 2 above)
Comment (javsalgar) [reply stable/mariadb/README.md]: Fixed

File: stable/mariadb/README.md (hunk 5 above)
Comment (javsalgar) [reply stable/mariadb/README.md]: Fixed

File: stable/mariadb/README.md (hunk 6 above)
Comment (javsalgar) [reply stable/mariadb/README.md]: Fixed

File: stable/mariadb/templates/NOTES.txt (hunk 7 above)
Comment (javsalgar) [reply stable/mariadb/templates/NOTES.txt]: I will put the release label, it will be more exact

File: stable/mariadb/values.yaml (hunk 9 above)
Comment (javsalgar) [reply stable/mariadb/values.yaml:231]: Added

File: stable/mariadb/README.md (hunk 4 above)
Comment (javsalgar) [reply stable/mariadb/README.md:85]: In terms of easiness, I think that this way is better for interacting with the command line. @prydonius @sameersbn thoughts?

**Comment by javsalgar** [comment]
> One thing that bugs me is that unless the bitnami image does some serious cool magic, you won't be able to get an additional slave up and running by simply increasing the number of replicas for the slaves. I think that should be very clearly explained in the README and probably the NOTES as well.

I tried it and it works AFAIK. The only requirement is that you use forcePassword and establish a password to be used, as stated in the Values.yml file:

```
  ##
  ## Option to force users to specify a password. That is required for 'helm upgrade' to work properly.
  ## If it is not force, a random password will be generated.
  forcePassword: false
```

Did you find any other issue when trying to scale?

**Comment by timstoop** [comment]
I didn't try it, but I don't see how it will copy the data from the master to the new slave? Or is this a feature in MariaDB that it can do this automatically? (I'm used to MySQL, so not aware of any kind of feature like this in MariaDB.)

**Comment by javsalgar** [comment]
@timstoop I just retried again and the data is indeed replicated. I created a table, scaled the slaves, entered one of the new slaves and the data is there:

```
javsalgar@javsalgar-MacBookPro:~/projects/javsalgar-charts/stable$ kubectl scale statefulsets alert-arachnid-mariadb-slave --replicas=4
statefulset "alert-arachnid-mariadb-slave" scaled

^Cjavsalgar@javsalgar-MacBookPro:~/projects/javsalgar-charts/stable$ kubectl exec -ti alert-arachnid-mariadb-slave-3 bash


I have no name!@alert-arachnid-mariadb-slave-3:/$ 
I have no name!@alert-arachnid-mariadb-slave-3:/$ mysql -uroot -p 
Enter password: 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 15
Server version: 10.2.14-MariaDB-log Source distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> use test;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [test]> select * from departments;
+---------+-----------+
| dept_no | dept_name |
+---------+-----------+
| a123    | Test Dpto |
+---------+-----------+
1 row in set (0.00 sec)
```

**Comment by prydonius** [comment]
@timstoop I believe this is using the binary log replication feature in MariaDB (https://mariadb.com/kb/en/library/replication-overview/) - not sure if this exists in MySQL. Is that correct @javsalgar?

File: stable/mariadb/README.md (hunk 4 above)
Comment (prydonius) [reply stable/mariadb/README.md:85]: Sounds like other charts set the entire affinity block so we should probably do that to stay consistent. Most likely, you're going to set those values in a values.yaml file anyway.

**Comment by javsalgar** [comment]
Yes, the bin-log properties are being set.

**Comment by timstoop** [comment]
Ok, but that assumes the binlog is still available. Old binlogs are automatically removed (after 7 days by default, I think, but you can set this in the config or even expire them manually). If there's no mechanism to actually sync based on the data (and vanilla MySQL doesn't have any and I couldn't find one for MariaDB on a quick search), this will break once the binlogs start rotating. It's also no option to keep the binlogs around, as they become large very quickly. Could you perform `PURGE BINARY LOGS` on the master and see what happens after you scale then? I expect it to fail... If you're unable to, I'll deploy and test it myself this evening/tomorrow.

(https://mariadb.com/kb/en/library/sql-commands-purge-logs/ docs about the command)

**Comment by javsalgar** [comment]
This is what I did.

- I scaled down to 1

```
javsalgar@javsalgar-MacBookPro:/tmp$ kubectl scale statefulsets alert-arachnid-mariadb-slave --replicas=1
statefulset "alert-arachnid-mariadb-slave" scaled
```

Master:

```
MariaDB [test]> insert into departments values('Test', 'Test');
Query OK, 1 row affected (0.01 sec)

MariaDB [test]> PURGE BINARY LOGS BEFORE '2018-04-20';
Query OK, 0 rows affected (0.00 sec)

MariaDB [test]> Bye
```

I scaled the them up again:

```
javsalgar@javsalgar-MacBookPro:/tmp$ kubectl scale statefulsets alert-arachnid-mariadb-slave --replicas=3
statefulset "alert-arachnid-mariadb-slave" scaled
```

- Checked if the new row is there

```
javsalgar@javsalgar-MacBookPro:/tmp$ kubectl exec -ti alert-arachnid-mariadb-slave-2 bash
I have no name!@alert-arachnid-mariadb-slave-2:/$ mysql -uroot -p
Enter password: 
Welcome to the MariaDB monitor.  Commands end with ; or \g.
Your MariaDB connection id is 50
Server version: 10.2.14-MariaDB-log Source distribution

Copyright (c) 2000, 2018, Oracle, MariaDB Corporation Ab and others.

Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.

MariaDB [(none)]> 
MariaDB [(none)]> use test;
Reading table information for completion of table and column names
You can turn off this feature to get a quicker startup with -A

Database changed
MariaDB [test]> select * from departments;
+---------+-----------+
| dept_no | dept_name |
+---------+-----------+
| Test    | Test      |
| a123    | Test Dpto |
+---------+-----------+
```

It seems to be there, is there anything I am missing?

**Comment by prydonius** [comment]
Looks like:
```
MariaDB [test]> PURGE BINARY LOGS BEFORE '2018-04-20';
Query OK, 0 rows affected (0.00 sec)
```

didn't actually purge anything @javsalgar, maybe remove the BEFORE condition in case that's incorrect?

**Comment by timstoop** [comment]
From the docs, it clearly states it needs to import a backup from the master: "Now, with the lock still in place, copy the data from the master to the slave. See Backup, Restore and Import for details on how to do this." (https://mariadb.com/kb/en/library/setting-up-replication/). I'm curious how this step is handled here. I'll have some time this evening to test this myself, but I'm pretty positive this cannot work, although I have no explanation for why your test worked, aside that maybe, as @prydonius says, the binlogs are not actually removed. You can check this easily in the master container itself, there should no longer be any bin-log files (or just a single, very small one) after you run the purge command.

For me, this is not a blocking problem for the PR, but we need to make sure and tell the end user what to expect here and document it. I have extensive experience with MySQL and setting up slaves has always been a chore. I'd welcome a truly easy fix for it, so if that's provided by this, I'll have to send someone their favourite beverage to thank them for making this work so easy. But I really doubt it works... :(

**Comment by prydonius** [comment]
> For me, this is not a blocking problem for the PR, but we need to make sure and tell the end user what to expect here and document it.

+1 completely agree, the downsides to the approach should be clearly stated as this will help users decide between alternative setups (e.g. Galera).

**Comment by timstoop** [comment]
They actually fixed it at Bitnami: `Use Bitnami’s replicated configuration, which uses the native cloud provider APIs to replicate the database from the master node to a configurable number of replicas.` (source: https://bitnami.com/stack/mariadb) that is... pretty awesome actually!

**Comment by timstoop** [comment]
I'd still like to see this pointed out in the README, however, as people cannot just substitute other images. (But why would you really want to if you get this feature??)

**Comment by prydonius** [comment]
> I'd still like to see this pointed out in the README

Also given this won't work on baremetal clusters?

**Comment by timstoop** [comment]
I don't know if it won't work on baremetal, I think it depends on the storage backend, but I'm unable to find the scripts used.

File: stable/mariadb/README.md (hunk 4 above)
Comment (javsalgar) [reply stable/mariadb/README.md:85]: I see that the antiaffinity block we currently set depends on template values, so it could not be put directly in the Values.yaml file. Thoughts?

Comment (javsalgar) [reply stable/mariadb/README.md:85]: ping @prydonius @timstoop

File: stable/mariadb/templates/_helpers.tpl [hunk 10]
```diff
@@ -9,24 +9,38 @@ Expand the name of the chart.
...
-Create chart name and version as used by the chart label.
-*/}}
 {{- define "mariadb.chart" -}}
 {{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" -}}
 {{- end -}}
+
+{{/*
+Return the proper MariaDB image name
+*/}}
+{{- define "mariadb.image" -}}
```
Comment (prydonius) [inline stable/mariadb/templates/_helpers.tpl]: Can we fix these? See https://github.com/kubernetes/charts/blob/master/stable/redis/templates/_helpers.tpl#L48

**Review by prydonius — CHANGES_REQUESTED** [review]

File: stable/mariadb/templates/_helpers.tpl (hunk 10 above)
Comment (javsalgar) [reply stable/mariadb/templates/_helpers.tpl]: Done

File: stable/mariadb/values.yaml [hunk 11]
```diff
@@ -1,110 +1,233 @@
...
   registry: docker.io
   repository: bitnami/mariadb
-  tag: 10.1.32
+  tag: 10.2.14
   ## Specify a imagePullPolicy
   ## Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
   ## ref: http://kubernetes.io/docs/user-guide/images/#pre-pulling-images
   ##
-  pullPolicy: IfNotPresent
+  pullPolicy: Always
```
Comment (prydonius) [inline stable/mariadb/values.yaml]: please set this back to IfNotPresent

File: stable/mariadb/values.yaml [hunk 12]
```diff
@@ -1,110 +1,233 @@
...
-
-
 service:
   ## Kubernetes service type
   type: ClusterIP
+  port: 3306
 
-  ## Specify the nodePort value for the LoadBalancer and NodePort service types.
-  ## ref: https://kubernetes.io/docs/concepts/services-networking/service/#type-nodeport
+root:
```
Comment (prydonius) [inline stable/mariadb/values.yaml]: this is misleading, should it be `rootUser`?

**Review by prydonius — CHANGES_REQUESTED** [review]

File: stable/mariadb/values-production.yaml [hunk 13]
```diff
@@ -22,7 +22,7 @@ service:
   type: ClusterIP
   port: 3306
 
-root:
+rootValues:
```
Comment (prydonius) [inline stable/mariadb/values-production.yaml]: should be rootUser

**Review by prydonius — APPROVED** [review]
